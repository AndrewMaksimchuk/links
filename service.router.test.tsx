// @ts-nocheck
import { describe, test, expect, spyOn } from 'bun:test'
import { HonoRequest } from './node_modules/hono/dist/request'
import { Context } from './node_modules/hono/dist/context'
import { Router } from './service.router'
import { Links } from "./component.links"
import { testUserLinks } from './test.fixture'


let service: Router


describe('Service Router', () => {
    test('should have one or more routes', () => {
        service = new Router()
        const length = Object.keys(service.routes).length
        expect(length).toBeGreaterThan(0)
    })


    test('usePrivateRoute(), should pass to private routes', async () => {
        const req = new Request('http://localhost/')
        const ctx = new Context(req)
        const service = new Router(
            {
                findByPasswordHash: () => ({}),
            },
            {
                getLoginToken: async () => '',
            }
        )
        const honoNext = {
            next: () => ({ status: 200 })
        }
        const res = await service.usePrivateRoute(ctx, honoNext.next)
        expect(res.status).toBe(200)
    })


    test('usePrivateRoute(), should not pass to private routes', async () => {
        const req = new Request('http://localhost/')
        const ctx = new Context(req)
        const service = new Router(
            {
                findByPasswordHash: () => null,
            },
            {
                getLoginToken: async () => false,
            }
        )
        const res = await service.usePrivateRoute(ctx)
        expect(res.status).toBe(302)
    })


    test('should redirect to dashboard page', async () => {
        const req = new Request('http://localhost/')
        const ctx = new Context(req)
        const service = new Router(
            {
                findByPasswordHash: () => ({}),
            },
            {
                getLoginToken: async () => '',
            }
        )
        const { status } = await service.main(ctx)
        expect(status).toBe(302)
    })


    test('should not redirect to dashboard page, but show main page', async () => {
        const req = new Request('http://localhost/')
        const ctx = new Context(req)
        const service = new Router(
            {
                findByPasswordHash: () => ({}),
            },
            {
                getLoginToken: async () => undefined,
            }
        )
        const res = await service.main(ctx)
        expect(res.status).toBe(200)
        expect(await Bun.readableStreamToText(res.body)).toMatchSnapshot()
    })


    test('should not show dashboard, but redirect to login page, can`t get user', async () => {
        const req = new Request('http://localhost/dashboard')
        const ctx = new Context(req)
        const service = new Router(
            {
                findByPasswordHash: () => ({}),
            },
            {
                getLoginToken: async () => undefined,
            }
        )
        const { status } = await service.dashboard(ctx)
        expect(status).toBe(302)
    })



    test('should show dashboard', async () => {
        const req = new Request('http://localhost/dashboard')
        const ctx = new Context(req)
        const service = new Router(
            {
                findByPasswordHash: () => ({
                    user_id: 1,
                }),
            },
            {
                getLoginToken: async () => '',
            },
            {
                getLinks: () => [],
            },
            {
                getTags: () => [],
            }, {}, {},
            {
                getLinkView: () => [<Links view='card' />, 'card']
            }
        )
        const { status, body } = await service.dashboard(ctx)
        expect(status).toBe(200)
        expect(await Bun.readableStreamToText(body)).toMatchSnapshot()
    })


    test('search, should ask to write more letters', async () => {
        const client = new Request('http://localhost/search', { method: "POST", body: new URLSearchParams("search=a") })
        const req = new HonoRequest(client)
        const ctx = new Context(req)
        const service = new Router()
        const res = await service.search(ctx)
        expect(res.headers.get('Content-Type')).toMatch('text/html')
        expect(await res.text()).toBe("Need more letters!")
    })


    test('search, should not find a user', async () => {
        const client = new Request('http://localhost/search', { method: "POST", body: new URLSearchParams("search=abc") })
        const req = new HonoRequest(client)
        const ctx = new Context(req)
        const service = new Router({},
            {
                getLoginToken: () => undefined
            }
        )
        const res = await service.search(ctx)
        expect(res.headers.get('Content-Type')).toMatch('text/html')
        expect(await res.text()).toBe("Can`t find you!")
    })


    test('search, should not find any results', async () => {
        const client = new Request('http://localhost/search', { method: "POST", body: new URLSearchParams("search=abc") })
        const req = new HonoRequest(client)
        const ctx = new Context(req)
        const service = new Router(
            {
                getUserData: () => { }
            },
            {
                getLoginToken: () => ''
            }, {}, {},
            {
                search: () => []
            }
        )
        const res = await service.search(ctx)
        expect(res.headers.get('Content-Type')).toMatch('text/html')
        expect(await res.text()).toBe("Nothing found")
    })


    test('search, should find results', async () => {
        const client = new Request('http://localhost/search', { method: "POST", body: new URLSearchParams("search=T62") })
        const req = new HonoRequest(client)
        const ctx = new Context(req)
        const service = new Router(
            {
                getUserData: () => { }
            },
            {
                getLoginToken: () => ''
            }, {}, {},
            {
                search: () => testUserLinks
            }, {},
            {
                getLinkView: () => [<Links view='card' />]
            }
        )
        const res = await service.search(ctx)
        expect(res.headers.get('Content-Type')).toMatch('text/html')
        expect(await Bun.readableStreamToText(res.body)).toMatchSnapshot()
    })


    test('paginationViewUpdate(), should show notification with "warn" status and message', async () => {
        const client = new Request('http://localhost/search', { method: "POST", body: new URLSearchParams("search=T62") })
        const req = new HonoRequest(client)
        const ctx = new Context(req)
        const service = new Router(
            {},
            {
                getLoginToken: () => undefined
            }
        )
        const res = await service.paginationViewUpdate(ctx)
        expect(res.headers.get('Content-Type')).toMatch('text/html')
        expect(await Bun.readableStreamToText(res.body)).toMatchSnapshot()
    })


    test('paginationViewUpdate(), should show view as card', async () => {
        const client = new Request('http://localhost/search', { method: "POST", body: new URLSearchParams("search=T62") })
        const req = new HonoRequest(client)
        const ctx = new Context(req)
        const service = new Router(
            {
                getUserData: () => { },
                findByPasswordHash: () => ({}),
            },
            {
                getLoginToken: () => ''
            }, {}, {},
            {
                search: () => testUserLinks
            },
            {
                setActivePage: () => { }
            },
            {
                getLinkView: () => [<Links view='card' />]
            }
        )
        const res = await service.paginationViewUpdate(ctx)
        expect(res.headers.get('Content-Type')).toMatch('text/html')
        expect(await Bun.readableStreamToText(res.body)).toMatchSnapshot()
    })


    test('paginationViewUpdate(), should show view as table', async () => {
        const client = new Request('http://localhost/search', { method: "POST", body: new URLSearchParams("search=T62") })
        const req = new HonoRequest(client)
        const ctx = new Context(req)
        const service = new Router(
            {
                getUserData: () => ({ user_id: 1 }),
                findByPasswordHash: () => ({}),
            },
            {
                getLoginToken: () => ''
            },
            {
                getLinks: () => []
            }, {},
            {
                search: () => testUserLinks
            },
            {
                setActivePage: () => { }
            },
            {
                getLinkView: () => [<Links view='table' />]
            }
        )
        const res = await service.paginationViewUpdate(ctx)
        expect(res.headers.get('Content-Type')).toMatch('text/html')
        expect(await Bun.readableStreamToText(res.body)).toMatchSnapshot()
    })


    test('settings(), should show dashboard', async () => {
        const req = new Request('http://localhost/dashboard')
        const ctx = new Context(req)
        const service = new Router(
            {
                findByPasswordHash: () => ({}),
            },
            {
                getLoginToken: async () => '',
            }, {}, {}, {}, {},
            {
                getLinkView: () => [<Links view='card' />, 'card']
            }
        )
        const { status, body } = await service.settings(ctx)
        expect(status).toBe(200)
        expect(await Bun.readableStreamToText(body)).toMatchSnapshot()
    })
})
