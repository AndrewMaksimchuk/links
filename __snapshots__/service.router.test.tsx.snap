// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`Service Router should not redirect to dashboard page, but show main page 1`] = `
"<html lang="en"><head><meta charset="UTF-8"/><meta name="viewport" content="width=device-width, initial-scale=1.0"/><link rel="icon" type="image/svg" href="favicon.svg"/><link rel="stylesheet" href="pico.yellow.min.css"/><script src="htmx.min.js" defer=""></script><script src="alpine.min.js" defer=""></script><title>Links</title></head><body class="container"><progress id="indicator" class="htmx-indicator" style="position: fixed; top:0; left: 0;"></progress><form hx-post="/login" hx-target="this" hx-swap="outerHTML" hx-push-url="true" id="form-entrance" class="form-entrance"><fieldset><legend style="font-size: 2rem; margin: auto;">Registration | Entrance</legend><label for="tel">Phone number</label><input type="tel" id="tel" name="telephone" title="Must containe only numbers" inputmode="tel" pattern="[0-9]{10,13}" autocomplete="off" minlength="10" maxlength="13" placeholder="0665243347" autofocus="" required=""/><label for="password">Password</label><input type="password" id="password" name="password" inputmode="text" autocomplete="current-password" minlength="10" title="Must be long, contain different symbols" placeholder="s8t!~RbR9i$^cK" required=""/><button type="submit">Log in</button></fieldset></form><footer></footer><section id="notification" style="--pico-background-color:var(--pico-form-element-background-color);--pico-border-color:var(--pico-form-element-border-color);--pico-color:var(--pico-form-element-color);--pico-box-shadow:none;position:fixed;max-width:300px;outline:0;top:var(--pico-form-element-spacing-vertical);right:var(--pico-form-element-spacing-horizontal);border:var(--pico-border-width) solid var(--pico-border-color);border-radius:var(--pico-border-radius);background-color:var(--pico-background-color);box-shadow:var(--pico-box-shadow);color:var(--pico-color);font-weight:var(--pico-font-weight);transition:background-color var(--pico-transition), border-color var(--pico-transition), color var(--pico-transition), box-shadow var(--pico-transition)"><script>
                const onNotify = new CustomEvent('notifyClose', { bubbles: true })
                document.body.addEventListener("notifyClose", function(event){
                    const selector = '[data-notificationid="' + event.detail.notificationid + '"]'
                    setTimeout(() => document.querySelectorAll(selector)[0]?.remove(), 3000)
                })
            </script></section></body></html>"
`;

exports[`Service Router should show dashboard 1`] = `
"<html lang="en"><head><meta charset="UTF-8"/><meta name="viewport" content="width=device-width, initial-scale=1.0"/><link rel="icon" type="image/svg" href="favicon.svg"/><link rel="stylesheet" href="pico.yellow.min.css"/><script src="htmx.min.js" defer=""></script><script src="alpine.min.js" defer=""></script><title>Links</title></head><body class="container"><progress id="indicator" class="htmx-indicator" style="position: fixed; top:0; left: 0;"></progress><header><div style="display:flex;justify-content:space-between;align-items:center;margin-bottom:var(--pico-typography-spacing-vertical)"><h1 id="userGreeting" style="margin: 0;"><a href="/dashboard">Hello, my friend!</a></h1><div style="display: flex; align-items: center; gap: 1rem;"><label><span>Dark</span><input type="checkbox" role="switch" style="margin-inline: .5em;" x-data="{
        dataTheme: undefined, 
        dataThemeValues: [&#39;dark&#39;, &#39;light&#39;],
        get dataThemeInputValue() {
            return &#39;dark&#39; === this.dataTheme ? false : true; 
        },
        setThemeCookie() {
            document.cookie = &#39;theme=&#39; + this.dataTheme + &#39;;SameSite=Strict;&#39;;
        },
        getThemeCookie() {
            const theme = document.cookie?.includes(&#39;theme&#39;) ? document.cookie.split(&#39;=&#39;).at(1) : undefined;
            return this.dataThemeValues.includes(theme) ? theme : undefined;
        },
        setTheme() {
            this.setThemeCookie();
            document.documentElement.setAttribute(&#39;data-theme&#39;, this.dataTheme);
        },
        toggleTheme() {
            this.dataTheme = &#39;dark&#39; === this.dataTheme ? &#39;light&#39; : &#39;dark&#39;;
            this.setTheme();
        },
            init() {
                const cookieTheme = this.getThemeCookie();
                if(cookieTheme) {
                    this.dataTheme = cookieTheme;
                    this.setTheme();
                    return;
                }
                this.dataTheme = window.matchMedia(&#39;(prefers-color-scheme: dark)&#39;).matches ? &#39;dark&#39; : &#39;light&#39;;
            },
    }" x-on:click="toggleTheme" x-model="dataThemeInputValue"/><span>Light</span></label><button class="outline secondary" onclick="settingsView.toggleAttribute(&#39;open&#39;); document.body.classList.toggle(&#39;modal-is-open&#39;);history.pushState({}, &#39;&#39;, &#39;/settings&#39;);">Settings</button><form action="/logout" method="post" style="margin: 0;"><button type="submit" class="outline" style="margin: 0;">Logout</button></form></div></div><nav style="display: grid; grid-template-columns: 1fr;"><form id="linkAddForm" hx-post="/link-add" hx-target="main" hx-swap="innerHTML" hx-on--after-request="this.reset()" hx-indicator="#indicator" hx-disabled-elt="#linkAddSubmitButton" hx-vals="js:{created_at: Date.now()}"><label for="link-add">New link</label><input type="url" name="url" id="link-add" placeholder="https://picocss.com/docs/forms" aria-describedby="link-add-aria-description" required="" minlength="12"/><small id="link-add-aria-description">Write ot paste link with https protocol</small><button id="linkAddSubmitButton" type="submit">Add new link</button></form><form hx-post="/search" hx-target="#searchResult" hx-swap="innerHTML" hx-on--before-request="searchClearButton?.remove()" hx-on--after-request="this.reset()" hx-indicator="#indicator" hx-disabled-elt="#searchFormSubmitButton" role="search"><input type="search" name="search" id="search" placeholder="picocss" minlength="2" required="" autofocus=""/><button id="searchFormSubmitButton" type="submit">Search</button></form><section id="searchResult"></section><div style="display: flex; align-items: center;"><fieldset><legend>Select view</legend><label><span>Card </span><input name="viewState" type="checkbox" role="switch" style="margin-inline: .5em;" hx-post="/link-change-view" hx-target="main" hx-swap="innerHTML"/><span>Table</span></label></fieldset><span id="linksCounter" hx-swap-oob="outerHTML" style="width: 10ch; text-align: right;">0 pcs</span></div></nav></header><dialog id="settingsView"><article><header style="display: flex; align-items: center;"><h2>Settings</h2><button class="outline" style="margin-left: auto;" onclick="settingsView.toggleAttribute(&#39;open&#39;); document.body.classList.toggle(&#39;modal-is-open&#39;);history.back();" hx-post="/link-add-form-update" hx-target="#linkAddForm" hx-swap="outerHTML">Close</button></header><div><form hx-post="/user-update-name" hx-target="this" hx-swap="outerHTML" hx-on--after-request="this.reset()"><label for="userName">Write your name</label><input type="text" id="userName" name="userName" inputmode="text" minlength="1" required="" value="" placeholder="Taras"/><input type="submit" value="Update name"/></form><a href="/user-data-download">Download your data</a></div><div><form hx-post="/tag-create" hx-target="#createTagResponse" hx-swap="afterbegin" hx-on--after-request="this.reset()"><fieldset><legend style="text-align: center; width: 100%; font-weight: bold;">Create new tag</legend><label for="inputTagName">Write tag name</label><input type="text" name="name" id="inputTagName" placeholder="web" inputmode="text" minlength="3" required="" autofocus=""/><label for="inputTagColor">Choose a color</label><input type="color" name="color" id="inputTagColor" required=""/><input type="submit" value="Create tag"/></fieldset></form><p id="createTagResponse" style="display: flex; flex-wrap: wrap; gap: 1rem; min-height: 2rem;"></p></div><footer></footer></article></dialog><main><section class="grid" style="grid-template-columns:repeat(4, 1fr)"></section></main><footer></footer><section id="notification" style="--pico-background-color:var(--pico-form-element-background-color);--pico-border-color:var(--pico-form-element-border-color);--pico-color:var(--pico-form-element-color);--pico-box-shadow:none;position:fixed;max-width:300px;outline:0;top:var(--pico-form-element-spacing-vertical);right:var(--pico-form-element-spacing-horizontal);border:var(--pico-border-width) solid var(--pico-border-color);border-radius:var(--pico-border-radius);background-color:var(--pico-background-color);box-shadow:var(--pico-box-shadow);color:var(--pico-color);font-weight:var(--pico-font-weight);transition:background-color var(--pico-transition), border-color var(--pico-transition), color var(--pico-transition), box-shadow var(--pico-transition)"><script>
                const onNotify = new CustomEvent('notifyClose', { bubbles: true })
                document.body.addEventListener("notifyClose", function(event){
                    const selector = '[data-notificationid="' + event.detail.notificationid + '"]'
                    setTimeout(() => document.querySelectorAll(selector)[0]?.remove(), 3000)
                })
            </script></section></body></html>"
`;

exports[`Service Router search, should find results 1`] = `"<section class="grid" style="grid-template-columns:repeat(4, 1fr)"><article id="l1" style="display:grid;grid-template-rows:115px 1fr auto"><header style="overflow:hidden;text-overflow:ellipsis;display:-webkit-box;-webkit-line-clamp:3;-webkit-box-orient:vertical" title="-">-</header><p style="overflow:hidden;text-overflow:ellipsis"><a href="web.dev" target="_blank" title="web.dev">web.dev</a></p><footer><p style="text-align: right;">Thu Jan 01 1970</p><p style="text-align: right;"><span style="--pico-color:var(--pico-switch-color);color:blue;border:solid 1px blue;border-radius:1.25em;background-color:blue30;line-height:1.25em;padding:0.25rem 0.75rem">web</span></p><p style="display:flex;flex-wrap:wrap;align-items:center;justify-content:right;gap:1rem"><button type="button" style="margin: 0; padding: 0.125rem 0.25rem;" class="outline" hx-get="/link-edit" hx-vals="{&quot;linkId&quot;: &quot;1&quot;}" hx-target="#l1" hx-swap="outerHTML">Edit</button><button type="button" style="margin: 0; padding: 0.125rem 0.25rem;" class="outline" hx-delete="/link-delete" hx-target="#notification" hx-swap="afterbegin" hx-on--after-request="this.parentElement.parentElement.parentElement.remove()" hx-indicator="#indicator" hx-disabled-elt="this" name="link_id" value="1">Delete</button></p></footer></article><article id="l3" style="display:grid;grid-template-rows:115px 1fr auto"><header style="overflow:hidden;text-overflow:ellipsis;display:-webkit-box;-webkit-line-clamp:3;-webkit-box-orient:vertical" title="-">-</header><p style="overflow:hidden;text-overflow:ellipsis"><a href="html.dev" target="_blank" title="html.dev">html.dev</a></p><footer><p style="text-align: right;">Thu Jan 01 1970</p><p style="text-align: right;"><span style="--pico-color:var(--pico-switch-color);color:blue;border:solid 1px blue;border-radius:1.25em;background-color:blue30;line-height:1.25em;padding:0.25rem 0.75rem">html</span></p><p style="display:flex;flex-wrap:wrap;align-items:center;justify-content:right;gap:1rem"><button type="button" style="margin: 0; padding: 0.125rem 0.25rem;" class="outline" hx-get="/link-edit" hx-vals="{&quot;linkId&quot;: &quot;3&quot;}" hx-target="#l3" hx-swap="outerHTML">Edit</button><button type="button" style="margin: 0; padding: 0.125rem 0.25rem;" class="outline" hx-delete="/link-delete" hx-target="#notification" hx-swap="afterbegin" hx-on--after-request="this.parentElement.parentElement.parentElement.remove()" hx-indicator="#indicator" hx-disabled-elt="this" name="link_id" value="3">Delete</button></p></footer></article><article id="l3" style="display:grid;grid-template-rows:115px 1fr auto"><header style="overflow:hidden;text-overflow:ellipsis;display:-webkit-box;-webkit-line-clamp:3;-webkit-box-orient:vertical" title="-">-</header><p style="overflow:hidden;text-overflow:ellipsis"><a href="css.dev" target="_blank" title="css.dev">css.dev</a></p><footer><p style="text-align: right;">Thu Jan 01 1970</p><p style="text-align: right;"><span style="--pico-color:var(--pico-switch-color);color:red;border:solid 1px red;border-radius:1.25em;background-color:red30;line-height:1.25em;padding:0.25rem 0.75rem">css</span></p><p style="display:flex;flex-wrap:wrap;align-items:center;justify-content:right;gap:1rem"><button type="button" style="margin: 0; padding: 0.125rem 0.25rem;" class="outline" hx-get="/link-edit" hx-vals="{&quot;linkId&quot;: &quot;3&quot;}" hx-target="#l3" hx-swap="outerHTML">Edit</button><button type="button" style="margin: 0; padding: 0.125rem 0.25rem;" class="outline" hx-delete="/link-delete" hx-target="#notification" hx-swap="afterbegin" hx-on--after-request="this.parentElement.parentElement.parentElement.remove()" hx-indicator="#indicator" hx-disabled-elt="this" name="link_id" value="3">Delete</button></p></footer></article></section><div id="searchResult" hx-swap-oob="beforebegin"><button id="searchClearButton" type="button" onclick="searchResult.innerText = &#39;&#39;; this.remove();">Clear search result view</button></div>"`;

exports[`Service Router paginationViewUpdate(), should show notification with "warn" status and message 1`] = `"<div id="notification" hx-swap-oob="afterbegin"><div style="padding:var(--pico-form-element-spacing-vertical) var(--pico-form-element-spacing-horizontal)"><h2 style="border-bottom:solid orange"></h2><p>Something is wrong!</p><button onclick="this.parentElement.remove()">Close</button></div></div>"`;

exports[`Service Router paginationViewUpdate(), should show view as card 1`] = `"<span id="linksCounter" hx-swap-oob="outerHTML" style="width: 10ch; text-align: right;">0 pcs</span><section class="grid" style="grid-template-columns:repeat(4, 1fr)"></section>"`;

exports[`Service Router paginationViewUpdate(), should show view as table 1`] = `"<span id="linksCounter" hx-swap-oob="outerHTML" style="width: 10ch; text-align: right;">0 pcs</span><table class="striped"><thead><tr><th scope="col">Title</th><th scope="col">Url</th><th scope="col">Date</th><th scope="col">Tags</th><th scope="col">Control</th></tr></thead><tbody></tbody></table>"`;

exports[`Service Router settings(), should show dashboard 1`] = `
"<html lang="en"><head><meta charset="UTF-8"/><meta name="viewport" content="width=device-width, initial-scale=1.0"/><link rel="icon" type="image/svg" href="favicon.svg"/><link rel="stylesheet" href="pico.yellow.min.css"/><script src="htmx.min.js" defer=""></script><script src="alpine.min.js" defer=""></script><title>Links</title></head><body class="container"><progress id="indicator" class="htmx-indicator" style="position: fixed; top:0; left: 0;"></progress><header><div style="display:flex;justify-content:space-between;align-items:center;margin-bottom:var(--pico-typography-spacing-vertical)"><h1 id="userGreeting" style="margin: 0;"><a href="/dashboard">Hello, my friend!</a></h1><div style="display: flex; align-items: center; gap: 1rem;"><label><span>Dark</span><input type="checkbox" role="switch" style="margin-inline: .5em;" x-data="{
        dataTheme: undefined, 
        dataThemeValues: [&#39;dark&#39;, &#39;light&#39;],
        get dataThemeInputValue() {
            return &#39;dark&#39; === this.dataTheme ? false : true; 
        },
        setThemeCookie() {
            document.cookie = &#39;theme=&#39; + this.dataTheme + &#39;;SameSite=Strict;&#39;;
        },
        getThemeCookie() {
            const theme = document.cookie?.includes(&#39;theme&#39;) ? document.cookie.split(&#39;=&#39;).at(1) : undefined;
            return this.dataThemeValues.includes(theme) ? theme : undefined;
        },
        setTheme() {
            this.setThemeCookie();
            document.documentElement.setAttribute(&#39;data-theme&#39;, this.dataTheme);
        },
        toggleTheme() {
            this.dataTheme = &#39;dark&#39; === this.dataTheme ? &#39;light&#39; : &#39;dark&#39;;
            this.setTheme();
        },
            init() {
                const cookieTheme = this.getThemeCookie();
                if(cookieTheme) {
                    this.dataTheme = cookieTheme;
                    this.setTheme();
                    return;
                }
                this.dataTheme = window.matchMedia(&#39;(prefers-color-scheme: dark)&#39;).matches ? &#39;dark&#39; : &#39;light&#39;;
            },
    }" x-on:click="toggleTheme" x-model="dataThemeInputValue"/><span>Light</span></label><button class="outline secondary" onclick="settingsView.toggleAttribute(&#39;open&#39;); document.body.classList.toggle(&#39;modal-is-open&#39;);history.pushState({}, &#39;&#39;, &#39;/settings&#39;);">Settings</button><form action="/logout" method="post" style="margin: 0;"><button type="submit" class="outline" style="margin: 0;">Logout</button></form></div></div><nav style="display: grid; grid-template-columns: 1fr;"><form id="linkAddForm" hx-post="/link-add" hx-target="main" hx-swap="innerHTML" hx-on--after-request="this.reset()" hx-indicator="#indicator" hx-disabled-elt="#linkAddSubmitButton" hx-vals="js:{created_at: Date.now()}"><label for="link-add">New link</label><input type="url" name="url" id="link-add" placeholder="https://picocss.com/docs/forms" aria-describedby="link-add-aria-description" required="" minlength="12"/><small id="link-add-aria-description">Write ot paste link with https protocol</small><button id="linkAddSubmitButton" type="submit">Add new link</button></form><form hx-post="/search" hx-target="#searchResult" hx-swap="innerHTML" hx-on--before-request="searchClearButton?.remove()" hx-on--after-request="this.reset()" hx-indicator="#indicator" hx-disabled-elt="#searchFormSubmitButton" role="search"><input type="search" name="search" id="search" placeholder="picocss" minlength="2" required="" autofocus=""/><button id="searchFormSubmitButton" type="submit">Search</button></form><section id="searchResult"></section><div style="display: flex; align-items: center;"><fieldset><legend>Select view</legend><label><span>Card </span><input name="viewState" type="checkbox" role="switch" style="margin-inline: .5em;" hx-post="/link-change-view" hx-target="main" hx-swap="innerHTML"/><span>Table</span></label></fieldset><span id="linksCounter" hx-swap-oob="outerHTML" style="width: 10ch; text-align: right;">0 pcs</span></div></nav></header><dialog id="settingsView"><article><header style="display: flex; align-items: center;"><h2>Settings</h2><button class="outline" style="margin-left: auto;" onclick="settingsView.toggleAttribute(&#39;open&#39;); document.body.classList.toggle(&#39;modal-is-open&#39;);history.back();" hx-post="/link-add-form-update" hx-target="#linkAddForm" hx-swap="outerHTML">Close</button></header><div><form hx-post="/user-update-name" hx-target="this" hx-swap="outerHTML" hx-on--after-request="this.reset()"><label for="userName">Write your name</label><input type="text" id="userName" name="userName" inputmode="text" minlength="1" required="" value="" placeholder="Taras"/><input type="submit" value="Update name"/></form><a href="/user-data-download">Download your data</a></div><div><form hx-post="/tag-create" hx-target="#createTagResponse" hx-swap="afterbegin" hx-on--after-request="this.reset()"><fieldset><legend style="text-align: center; width: 100%; font-weight: bold;">Create new tag</legend><label for="inputTagName">Write tag name</label><input type="text" name="name" id="inputTagName" placeholder="web" inputmode="text" minlength="3" required="" autofocus=""/><label for="inputTagColor">Choose a color</label><input type="color" name="color" id="inputTagColor" required=""/><input type="submit" value="Create tag"/></fieldset></form><p id="createTagResponse" style="display: flex; flex-wrap: wrap; gap: 1rem; min-height: 2rem;"></p></div><footer></footer></article></dialog><main><section class="grid" style="grid-template-columns:repeat(4, 1fr)"></section></main><footer></footer><section id="notification" style="--pico-background-color:var(--pico-form-element-background-color);--pico-border-color:var(--pico-form-element-border-color);--pico-color:var(--pico-form-element-color);--pico-box-shadow:none;position:fixed;max-width:300px;outline:0;top:var(--pico-form-element-spacing-vertical);right:var(--pico-form-element-spacing-horizontal);border:var(--pico-border-width) solid var(--pico-border-color);border-radius:var(--pico-border-radius);background-color:var(--pico-background-color);box-shadow:var(--pico-box-shadow);color:var(--pico-color);font-weight:var(--pico-font-weight);transition:background-color var(--pico-transition), border-color var(--pico-transition), color var(--pico-transition), box-shadow var(--pico-transition)"><script>
                const onNotify = new CustomEvent('notifyClose', { bubbles: true })
                document.body.addEventListener("notifyClose", function(event){
                    const selector = '[data-notificationid="' + event.detail.notificationid + '"]'
                    setTimeout(() => document.querySelectorAll(selector)[0]?.remove(), 3000)
                })
            </script></section></body></html>"
`;
